nomad/fsm_test.go

```diff
@@ -2836,6 +2848,183 @@ func TestFSM_ReconcileSummaries(t *testing.T) {
+	}
+
+	// Bump job to 1 and add a couple allocs
+
+	// alloc.ClientStatus = structs.AllocClientStatusRunning
+
+	// Create a new alloc here, have it be running
+
+	job2 := job1.Copy()
+	alloc2 := mock.Alloc()
+	alloc2.DesiredStatus = structs.AllocClientStatusRunning
+	alloc2.NodeID = node.ID
+	alloc2.Job = job2
+	alloc2.JobID = job2.ID
+	alloc2.ClientStatus = structs.AllocClientStatusRunning
+
+	// Create a new alloc here, have it be running
+	alloc3 := mock.Alloc()
+	alloc3.NodeID = node.ID
+	alloc3.Job = job2
+	alloc3.JobID = job2.ID
+	alloc3.ClientStatus = structs.AllocClientStatusRunning
+	require.NoError(t, state.UpsertJob(structs.MsgTypeTestSetup, 1020, nil, job2))
+	require.NoError(t, state.UpsertAllocs(structs.MsgTypeTestSetup, 1021, []*structs.Allocation{alloc2, alloc3}))
+
+	require.NoError(t, state.DeleteJobSummary(1030, job1.Namespace, job1.ID))
+
+	// Apply ReconcileJobSummaries request
+	resp = fsm.Apply(makeLog(buf))
+	if resp != nil {
+		t.Fatalf("resp: %v", resp)
+	}
+
+	out2, err := state.JobSummaryByID(ws, job1.Namespace, job1.ID)
+	require.NoError(t, err)
+
+	expected = structs.JobSummary{
+		JobID:     alloc1.Job.ID,
+		Namespace: alloc1.Job.Namespace,
+		Summary: map[string]structs.TaskGroupSummary{
+			"web": {
+				Queued:  8,
+				Failed:  1,
+				Running: 2,
+			},
+		},
+		VersionedSummary: map[uint64]structs.VersionedSummary{
+			// TODO: Why doesn't version 0 show up here?
+			0: {
+				Version: 0,
+				Groups: map[string]structs.TaskGroupSummary{
+					"web": {
+						Failed: 1,
+					},
+				},
+			},
+			1: {
+				Version: 1,
+				Groups: map[string]structs.TaskGroupSummary{
+					"web": {
+						Running: 2,
+					},
+				},
+			},
+		},
+		CreateIndex: 1000,
+		ModifyIndex: out2.ModifyIndex,
+	}
+
+	must.Eq(t, &expected, out2)
+
+	// Now let's GC an alloc and see how our summary updates
+	alloc3.ClientStatus = structs.AllocClientStatusFailed
+	require.NoError(t, state.UpsertAllocs(structs.MsgTypeTestSetup, 1040, []*structs.Allocation{alloc3}))
+
+	// Make the GC request
+	s1, cleanupS1 := TestServer(t, nil)
+	defer cleanupS1()
+	codec := rpcClient(t, s1)
+	gcreq := &structs.GenericRequest{
+		QueryOptions: structs.QueryOptions{
+			Region: "global",
+		},
+	}
+	var gcresp structs.GenericResponse
+	if err := msgpackrpc.CallWithCodec(codec, "System.GarbageCollect", gcreq, &gcresp); err != nil {
+		t.Fatalf("expect err")
+	}
+
+	testutil.WaitForResult(func() (bool, error) {
+		// Does the alloc still exist?
+		alloc2result, _ := state.AllocByID(ws, alloc2.ID)
+		logger.Info("=======checking on result", "alloc2", alloc2result)
+
+		alloc3result, err := state.AllocByID(ws, alloc3.ID)
+		logger.Info("=======checking on result", "alloc3", alloc3result)
+		if err != nil {
+			return false, err
+		}
+		if alloc3result != nil {
+			return false, fmt.Errorf("alloc %+v wasn't garbage collected", alloc3)
+		}
+
+		return true, nil
+	}, func(err error) {
+		t.Fatalf("err: %s", err)
+	})
```

Comment:
This part of the test is not working right now; I've decided to get eyes on the PR anyway. I'd like to have this verify that the alloc got GC'd and for it to show up in the summary regardless.

Reply:
The logic around what gets GC'd is pretty complex and not super relevant to what you're trying to test here. If you look at `nomad/core_sched.go`, you can find that ultimately the GC is going to call `Eval.Reap` with the alloc ID in its request. So you could call that directly, or even go one level down and call `state.DeleteEval` because that's the only logic you'll find in the `Eval.Reap` RPC's write.

At the very least, looking `state.DeleteEval` might help you figure out why the test is failing.